File: progs/python/race-n.py
import asyncio
import time

async def id(n):
    return n

async def forever():
    while True:
        await asyncio.sleep(0)

async def race_n(futs):
    # NOTE, no cancellation but the program will terminate
    for fut in asyncio.as_completed(futs):
        result = await fut
        if result is not None:
            return result
    return None

async def main():
    n = 42
    timer = time.perf_counter()
    futs = (id(i) if i == n else forever() for i in range(1001))
    res = await race_n(futs)
    assert res == n
    print(f"done {int((time.perf_counter() - timer) * 1_000_000)}μs")

if __name__ == "__main__":
    asyncio.run(main())

File: progs/javascript/misc/race-n.js
const id = async (n) => {
  return n;
};

const forever = async (signal) => {
  while (signal.aborted === false) {
    await new Promise(resolve => setTimeout(resolve, 50));
  }
  return;
};

const raceN = async (futs, controller) => {
  const res = await Promise.any(futs);
  // Must abort, otherwise program doesn't terminate
  controller.abort();
  return res;
};

const main = async () => {
  const abort = new AbortController();
  const n = 42;
  const timer = process.hrtime.bigint();
  const futs = Array.from({ length: 1001 }, (_, i) => (i === n ? id(i) : forever(abort.signal)));
  const res = await raceN(futs, abort);
  console.assert(res === n, `Result ${res} does not match expected value`);
  const elapsed = Number(process.hrtime.bigint() - timer) / 1000;
  console.log(`done ${elapsed}μs`);
};

main().catch(console.error);
